File: src/main/java/not/here/ykis/eventhook/Registry.java
package not.here.ykis.eventhook;

import java.util.ArrayList;

import org.bukkit.plugin.Plugin;
import org.bukkit.event.Listener;

import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.event.Event;
import org.bukkit.event.EventPriority;
import org.bukkit.event.HandlerList;

import java.io.File;
import java.util.List;

/** The registry, as its name suggests, builds, keeps record of and deregisters individual {@link Athlete}s */
class Registry implements Listener {

  private Plugin plugin;
  private ArrayList<Athlete> theRegister;

  Registry(Plugin plugin) {
    this.plugin = plugin;
    this.theRegister = new ArrayList<>();
  }

  /**
   * Create {@link Athlete}s from a given {@link ConfigurationSection}
   *
   * the {@code applicationForms} should be a mapping from a {@code name}
   * to the applicable parts of the arguments of {@link Athlete#Athlete(Plugin, String, Class, EventPriority, String, File, boolean, List) Athlete#Athlete}
   *
   * @param applicationForms the relevant section of configuration
   * @return total number of entries parsed and constructed
   */
  int processApplications(ConfigurationSection applicationForms) {
    for(String applicantName : applicationForms.getKeys(false)) {
      ConfigurationSection application = applicationForms.getConfigurationSection(applicantName);
      ConfigHelper reader = new ConfigHelper(this.plugin, application, applicantName);

      final Class<? extends Event> eventClass;
      final EventPriority eventPriority;
      final String execPath;
      final File workDir;
      final boolean announce;
      final List<String> execArgs;

      try {
        eventClass = reader.getEvent();
        eventPriority = reader.getPriority();
        execPath = reader.getExecPath();
        workDir = reader.getWorkDir();
        announce = reader.getAnnounce();
        execArgs = reader.getArgs();
      } catch (RuntimeException e) {
        this.plugin.getLogger().severe(String.format(
          "In config.yml[%s.%s]: %s",
          Constants.RUNNERS_CONTAINER, applicantName, e.getMessage()
        ));
        this.plugin.getLogger().warning("Entry Skipped");
        continue;
      }

      this.theRegister.add(new Athlete(this.plugin, applicantName, eventClass, eventPriority, execPath, workDir, announce, execArgs));
    }

    return this.theRegister.size();
  }

  /** Make each of the recorded {@link Athlete}s ready and register themselves to their corresponding {@link Event}s*/
  void makeReady() {
    this.theRegister.forEach(athlete -> athlete.onMyMark(this));
  }

  /** Centrally Unregisters all of the {@link Athlete}s from all of the {@link Event}s */
  void dismissParticipants() {
    HandlerList.unregisterAll(this);
  }
}

File: src/main/kotlin/Registry.kt
package not.here.ykis.eventhook

import java.io.File
import kotlin.script.experimental.api.ResultValue
import kotlin.script.experimental.jvmhost.BasicJvmScriptingHost

import org.bukkit.plugin.Plugin
import org.bukkit.event.Event
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener

/** The registry, as its name suggests, builds, keeps record of and (de)registers individual [Athlete]s  */
class Registry(private val plugin: Plugin) : Listener {

  private var theRegister: MutableSet<Athlete<*>> = mutableSetOf()
  val isLoaded: Boolean
    get() = this.theRegister.isNotEmpty()

  /**
   * Create [Athlete]s from the specified .kts script file
   *
   * @return total number of entries parsed and constructed
   */
  fun processApplications(): Int {
    val sourceFile = File(this.plugin.dataFolder, Constants.NAME_KTSFILE)
    val host = BasicJvmScriptingHost()
    val config = ScriptProxyConfig(host, returns = mutableSetOf())
    val proxy = ScriptingProxy(this.plugin.logger, sourceFile, config)

    with(this.plugin.logger){
      when(val result = proxy.evalFile()) {
        is ResultValue.Value -> {
          val value = result.value
          warning("Script %s returned a %s: %s".format(
              sourceFile.name, value?.javaClass?.name ?: "null", value.toString()
          ))
          // TODO Maybe further handling here?
        }
        is ResultValue.Unit ->
          info("Script %s finished".format(sourceFile.name))
        is ResultValue.Error -> {
          severe("Script %s thrown an Exception:".format(sourceFile.name))
          severe(result.error.stackTraceToString())
          return -1
        }
        ResultValue.NotEvaluated ->
          warning("Script %s did not ran".format(sourceFile.name))
        null -> {
          severe("Script %s failed to run".format(sourceFile.name))
          return -1
        }
      }
    }

    this.theRegister = config.returns.mapTo(mutableSetOf()) {
      Athlete(plugin, it)
    }

    return this.theRegister.size
  }

  /** Make each of the recorded [Athlete]s ready and register themselves to their corresponding [Event]s */
  fun makeReady() = theRegister.forEach { athlete -> athlete.onMyMark(this) }

  /** Centrally Unregisters all of the [Athlete]s from all of the [Event]s  */
  fun dismissParticipants() {
    HandlerList.unregisterAll(this)
    this.theRegister.clear()
  }

}
